src/tabbar.cpp

```diff
@@ -254,3 +279,65 @@ void TabBar::mousePressEvent(QMouseEvent *event)
        QTabBar::mousePressEvent(event);
     }
 }
+
+void TabBar::paintEvent(QPaintEvent *e)
+{
+    // Please keep it in sync with resources/css/style.css
+    // QTabBar::tab:selected { background-color: <value> }
+    const QColor selected_tab_bg_color = QColor(Qt::white);
+
+    // first, let Qt draw QTabBar normally
+    QTabBar::paintEvent(e);
+
+    // Then apply fade-out effect for long tab title on top:
+    QPainter p(this);
+
+    for (int i = 0; i < count(); ++i) {
+        bool need_fade_out = tabData(i).toBool();
+        if (! need_fade_out)
+            continue;
+
+        QStyleOptionTab tab;
+        initStyleOption(&tab, i);
+
+        QRect textRect = style()->subElementRect(QStyle::SE_TabBarTabText, &tab, this);
+
+        QRect tail = textRect;
+        int offset = textRect.width() * 0.8;
+        if (tab.direction == Qt::RightToLeft) {
+            tail.setWidth(textRect.width() - offset);
+        } else {
+            tail.setLeft(tail.x()+offset);
+            tail.setWidth(textRect.width() - offset);
```

Comment:
I would move the two `tail.setWidth(textRect.width() - offset);` out of the if/else.

Reply:
Rewritten without `offset` variable, not needed: `QRect::moveRight()` is really what I needed.